myhoard/web_server.py

```diff
@@ -78,6 +79,59 @@ async def backup_list(self, _request):
                     response["backups"] = self.controller.state["backups"]
                 return json_response(response)
 
+    async def backup_preserve(self, request):
+        with self._handle_request(name="backup_preserve"):
+            stream_id = request.match_info["stream_id"]
+            body = await self._get_request_json(request)
+            preserve_until = body.get("preserve_until")
+
+            if preserve_until is not None:
+                try:
+                    preserve_until = datetime.fromisoformat(preserve_until).astimezone(timezone.utc)
+                    now = datetime.now(timezone.utc)
+                    if preserve_until < now:
+                        raise BadRequest("`preserve_until` must be a date in the future.")
+                except ValueError:
+                    raise BadRequest("`preserve_until` must be a valid isoformat datetime string.")
+
+            self.controller.mark_backup_preservation(stream_id=stream_id, preserve_until=preserve_until)
+            wait_for_applied_preservation = body.get("wait_for_applied_preservation")
+            if wait_for_applied_preservation:
+                self.log.info(
+                    "Waiting up to %.1f seconds for preservation of backup %s to be applied.",
+                    wait_for_applied_preservation,
+                    stream_id,
+                )
+                start = time.monotonic()
+                while True:
+                    backup = self.controller.get_backup_by_stream_id(stream_id)
```

Comment:
Doesn't look like a public method named this would exist?

Reply:
apparently Pycharm's refactor played against.

Reply:
should be public now